services/stackexchange/stackexchange-base.spec.js

```diff
@@ -16,23 +31,47 @@ class DummyStackExchangeService extends StackExchangeBase {
   }
 }
 
-describe('StackExchangeBase', function () {
-  describe('auth', function () {
-    cleanUpNockAfterEach()
+// format is [class, example response from server]
+const testClasses = [
+  [DummyStackExchangeService, { message: 'fake message' }],
+  [StackExchangeMonthlyQuestions, { total: 8 }],
+  [StackExchangeReputation, { items: [{ reputation: 8 }] }],
+  [StackExchangeQuestions, { items: [{ count: 8 }] }],
+]
 
-    const config = { private: { stackapps_api_key: 'fake-key' } }
+for (const [serviceClass, dummyResponse] of testClasses) {
+  testAuth(serviceClass, dummyResponse)
+}
 
-    it('sends the auth information as configured', async function () {
-      const scope = nock('https://api.stackexchange.com')
-        .get('/2.2/tags/python/info')
-        .query({ key: 'fake-key' })
-        .reply(200, { message: 'fake message' })
+function testAuth(serviceClass, dummyResponse) {
+  describe(serviceClass.name, function () {
+    describe('auth', function () {
```

Comment:
This pattern of putting the `describe`s inside a function declaration is a bit unusual. It works, but for me it violates the [rule of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). More idiomatic would be to do it the other way round and  define a test helper, then write tests (at the module level) that use it. Something like https://github.com/badges/shields/blob/b2c2a1879616cce51cd562af2fac5f09592d5484/badge-maker/lib/make-badge.spec.js would be an example of that.

I guess the reason why we've done it this way round is because it becomes difficult to do `describe(serviceClass.name` the other way round.

I guess that leads me towards saying maybe the correct way to do this would be a test file for each service that does something more like

`stackexchange-monthlyquestions.spec.js`

```js
// bla bla imports
import { testAuth } from './test-helpers.js'

describe('StackExchangeMonthlyQuestions', function () {
  describe('auth', function () {
    it('sends the auth information as configured', async function () {
      testAuth(StackExchangeMonthlyQuestions, { total: 8 })
    })
  })
})
```

The other advantage of that is: If we add another service, it is more obvious that you also need to add tests while still abstracting the boilerplate. It is less obvious you would need to add a test case of a concrete implementation to `stackexchange-base.spec.js`.

How does that sound?

Reply:
I tried to make it more compact but this introduced some complexity and i also seem to result in a leas readable code. At least we learn from that attempt how to better move forward.

It does sound like a good idea to separate the files, with the scale of tests we use its better to keep it simple in a  single file for each auth test, similar to the example you provided here, rather then a more compact array in the base spec file, as there aren't that many.

I could rewrite this PR in a few days.